### PR TITLE
WIP: RooFit import/export of JSON/YAML files

### PR DESCRIPTION
**Pull request by cburgard**
The work of this MR has been presented in RooFit developer meetings a few times already. However, I would like to put this forward not as a "final MR" (would need to be rebased anyway), but as a way to get a preliminary review of this code.

Purpose: HistFactory and RooWorkspace should be able to import and export their contents from and to text files, using a common JSON/YAML syntax

Method: The code has been written in such a way as to

- Minimize the amount of changes to existing code
- Make it as easy as possible for users to add new serializers/deserializers for new or existing classes
- Provide a clean interface as much as possible.

Dependencies: The code requires a JSON/YML parser and emitter.
ROOT already includes the nlohmann JSON module, which is used as a default for the interface.
However, for development, I have been using the rapidyml package, which also supports reading and writing of YML files. As the code for this already exists, I have kept this in as an optional dependency. When available, it replaces the nlohmann interface and then also provides YML reading and writing capabilities, while the default is limited to JSON.

Please review:

- general coding style
- packaging: Should these new files continue to live in RooStats, or should they go to a new subpackage of RooFit?
- code placement: Is the way I split the functionality up into different source files acceptable, or would you wish for a smaller/higher granularity of the splitting?
- build system: Is the way I encoded the optional dependency acceptable?
- class and function names: Are the names of classes and functions acceptable, or should things be renamed to be more conformant?

Please don't review yet:
- lack of functionality - more things will be added over time
- individual violations of the style guide (e.g. missing spaces or newlines) 

Please specifically suggest where documenation should be added!
Does every function need a docstring, or just the ones announced in header files? Do these docstrings need to exist also for the Interface class added, and also for the implementations of that, or only for the RooFit-side of the code?

### CODE REVIEW

**Comment by lmoneta** [comment]
@phsft-bot build!

**Comment by hageboeck** [comment]
Hi Carsten,

here are some answers:

>  general coding style

I like the const-correctness fixes. Can you bundle them in an extra PR, and we merge them before the rest gets merged? This is a code improvement, anyway. The addition of functions such as `GetPreprocess...()` can probably added as well.

>    packaging: Should these new files continue to live in RooStats, or should they go to a new subpackage of RooFit?

All YML/JSON-related things would best go into something that can be switched on/off based on the availability of YML/JSON. Something like `roofit/YMLFactory`?
>    code placement: Is the way I split the functionality up into different source files acceptable, or would you wish for a smaller/higher granularity of the splitting?

I didn't really check that. Maybe we do this when there's a bit less noise after splitting off the const-ness fixes and interface changes.
>    build system: Is the way I encoded the optional dependency acceptable?

I don't know. @oshadura? @Axel-Naumann ?
>    class and function names: Are the names of classes and functions acceptable, or should things be renamed to be more conformant?

As long as classes and function names say what these things do, you have quite some freedom. Maybe just check whether the surrounding code capitalises functions or not.

> Please don't review yet:
> 
>    lack of functionality - more things will be added over time
>    individual violations of the style guide (e.g. missing spaces or newlines)

Sure, NP.

> Please specifically suggest where documenation should be added!
> Does every function need a docstring, or just the ones announced in header files? Do these docstrings need to exist also for the Interface class added, and also for the implementations of that, or only for the RooFit-side of the code?

It would be good if every function and class had documentation. They are best placed in the implementation like in
https://root.cern.ch/doc/master/RooAbsPdf_8cxx_source.html

Header-only functions can have small docstrings in the header. The class documentation should explain what the class does, maybe give examples of how to use it. The function documentation should at least say what the function does, what kind of inputs it needs, and what it outputs.
An a-bit-too-long example is:
https://root.cern.ch/doc/master/RooAbsPdf_8cxx_source.html#l01255

File: roofit/roostats/CMakeLists.txt [hunk 1]
```diff
@@ -141,3 +149,17 @@ ROOT_STANDARD_LIBRARY_PACKAGE(RooStats
...
 )
+
+target_include_directories(RooStats
+  PRIVATE
+  ${CMAKE_SOURCE_DIR}/io/io/res
+)
+
+
+if(yml)
+  FIND_PACKAGE(ryml)
```
Comment (oshadura) [inline roofit/roostats/CMakeLists.txt]: This should go to https://github.com/root-project/root/blob/master/cmake/modules/SearchInstalledSoftware.cmake#L255
Please take an example from xxhash as well as to add support of option fail-on-missing and possibly builtins in the future.

File: roofit/histfactory/CMakeLists.txt [hunk 2]
```diff
@@ -78,6 +81,14 @@ ROOT_STANDARD_LIBRARY_PACKAGE(HistFactory
     XMLParser
 )
 
+if(yml)
+  FIND_PACKAGE(ryml)
```
Comment (oshadura) [inline roofit/histfactory/CMakeLists.txt]: This should go to https://github.com/root-project/root/blob/master/cmake/modules/SearchInstalledSoftware.cmake#L255
Please take an example from xxhash as well as to add support of option fail-on-missing and possibly builtins in the future.

**Comment by cburgard** [comment]
Closing this MR, work continues [here](https://github.com/root-project/root/pull/5324)

**Pull request closed without merging**